### PR TITLE
Switch StreamField to use same panel component as other collapsible sections. Fix #9180, #8047

### LINKED ISSUE #9180: Accessibility - StreamField renders an empty heading (h3) element
### Issue Summary

* `StreamField` items render a block container with a `h3` element but this title will always be empty, providing no information to the user for the titles in this area of the UI.
* This blocks a critical AA accessibility criteria 2.4.6 Headings & Labels - https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels

### Steps to Reproduce

1. Open the bakerydemo
2. Open any page with a StreamField being used
3. Use an accessibility validator tool or use a screen reader to navigate to one inner stream field block
4. Expected: h3 should read out a title
5. Actual: h3 is empty

- I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (no)
- Found while investigating a solution for #8673 

### Screenshots

**See Firefox accessibility tool - the header under 'page body' is blank**

Screen Shot 2022-09-11 at 2 52 22 pm

**See NVDA - it reads the outer section's title 'Page body' but not the inner one 'paragraph block'**

Screen Shot 2022-09-11 at 2 50 55 pm

**See VoiceOver - does not read any heading for the paragraph block section**

> Left - inner paragraph block section, Right - outer page body section

Screen Shot 2022-09-11 at 2 49 49 pm

**See DOM inspector - H3 is empty**

Screen Shot 2022-09-11 at 2 43 08 pm

### Potential fix

https://github.com/wagtail/wagtail/blob/519caf25e9d088adb436cff66d4a82e5a35cbeed/client/src/components/StreamField/blocks/BaseSequenceBlock.js#L151-L154

My guess is that the lines above can merge into one, the div block__type and h3 header__title can be the same content, and the div can simply be removed.

This will come with some styling fixes to do also.

### Technical details

-   Python version: 3.10
-   Django version: 4.1
-   Wagtail version: 4.0 & also checked on latest master (4.1a)
-   Browser version: Tested on Firefox 104, Safari 15 on MacOS with VoiceOver, Edge 99 on Windows 10 with NVDA

### PR DESCRIPTION
**Pull request by thibaudcolas**
Like #9316, this is one of the initial tasks necessary for implementation of the new minimap component (#8673). Those changes are also relevant on their own, bridging the gap in UI behavior between StreamField and InlinePanel.

The high-level change is siwthcing StreamField blocks to use the same collapsible panel component as all top-level sections of the page, and InlinePanel children.

Detailed improvements that result from the above switch:

- Add ARIA `<section>` region for each StreamField block, with a heading, so screen reader users can more easily bypass blocks.
- Make sure panel headings and control buttons (for InlinePanel & StreamField only) display consistently regardless of the length of the text
- Add anchor links to StreamField, so users can navigate straight to a block.
- "Ctrl + f" search-in-page now reveals content within collapsed StreamField blocks (for browsers which support `hidden="until-found"`).

Bug fixes:

- StreamField blocks can now be collapsed with the keyboard
- StreamField block headings now have a label for screen reader users (were previously empty)

Less related StreamField changes which I made along the way to prepare for upcoming minimap:

- Add a "*" indicator for required StreamField blocks (single-field blocks, e.g. CharBlock).
- Change StreamField’s "block summary label" functionality so the content-derived label is always available for screen reader users. This helps make sure each section has a label that’s more likely to be unique.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
-   ~~[ ] For Python changes: Have you added tests to cover the new/fixed behaviour?~~
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**: Chrome 106, Firefox 104, Safari 15.6 on macOS 12.5.1
    -   [x] **Please list which assistive technologies [^3] you tested**: Safari VoiceOver, keyboard, WHCM
-   ~~[ ] For new features: Has the documentation been updated accordingly?~~

Aside from the addition of the "anchor" links and the "*" required indicators, StreamField should look the same as before. For InlinePanel, the only change is the removal of the space between blocks (UI inconsistency that wasn’t intended).

### CODE REVIEW

File: client/scss/components/_panel.scss [hunk 1]
```diff
@@ -45,6 +47,7 @@ $header-button-size: theme('spacing.8');
   align-content: center;
   color: theme('colors.primary.DEFAULT');
   border-radius: theme('borderRadius.full');
+  margin: 0;
```
Comment (thibaudcolas) [inline client/scss/components/_panel.scss:50]: This is what has slightly changed the UI for InlinePanel – so we display the buttons without any space in-between. Even though the icons are only 16x16px, the buttons are 32x32px, so this big enough to not require the extra space.

File: client/src/components/StreamField/StreamField.scss [hunk 2]
```diff
@@ -1,19 +1,28 @@
...
+  @apply w-sr-only;
+}
+
+// Only hide the block type if the field is collapsed *and* the title isn’t empty.
+[aria-expanded='false']
+  + .w-panel__heading
+  .c-sf-block__title:not(:empty)
+  + .c-sf-block__type {
+  display: none;
+}
```
Comment (thibaudcolas) [inline client/src/components/StreamField/StreamField.scss:22]: This selector is a bit frightening, but doing it like this in CSS saves quite a bit of JS.

File: client/src/components/StreamField/StreamField.scss [hunk 3]
```diff
@@ -1,19 +1,28 @@
...
+
+.c-sf-help .help {
+  // StreamField block help text is above the blocks rather than below.
+  margin-top: 0;
+  margin-bottom: theme('spacing.[1.5]');
+}
```
Comment (thibaudcolas) [inline client/src/components/StreamField/StreamField.scss:28]: This was the only code left in `c-sf-container`, so I’ve moved it here as-is.

File: client/src/components/StreamField/blocks/BaseSequenceBlock.js [hunk 4]
```diff
@@ -334,34 +359,7 @@ export class BaseSequenceChild extends EventEmitter {
...
 
   collapse() {
-    this.contentElement.hide().attr('aria-hidden', 'true');
-    const label = this.getTextLabel({ maxLength: 50 });
-    this.titleElement.text(label || '');
-    this.collapsed = true;
-    this.contentElement
-      .get(0)
-      .dispatchEvent(
-        new CustomEvent('commentAnchorVisibilityChange', { bubbles: true }),
```
Comment (thibaudcolas) [inline client/src/components/StreamField/blocks/BaseSequenceBlock.js:344]: This is now dispatched by the collapsible panels code, identically to what we’re doing here.

File: client/src/includes/panels.ts [hunk 5]
```diff
@@ -43,16 +44,23 @@ const toggleCollapsiblePanel = (
...
   const panel = toggle.closest<HTMLElement>('[data-panel]');
   const content = document.querySelector<HTMLDivElement>(
     `#${toggle.getAttribute('aria-controls')}`,
   );
 
-  if (!content || !panel) {
+  // Avoid initialising the same panel twice.
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  if (!content || !panel || panel.collapsibleInitialised) {
```
Comment (thibaudcolas) [inline client/src/includes/panels.ts:56]: Not my proudest work but this helps resolve double initialisation issues with this component. This is exactly the type of thing that won’t be an issue anymore once we switch over to Stimulus, which would be a great fit for this component, so I did the quick fix and moved on.

**Review by kaedroho — APPROVED** [review]
This looks good to me, thanks!

File: client/src/components/StreamField/StreamField.scss (hunk 2 above)
Comment (kaedroho) [reply client/src/components/StreamField/StreamField.scss:22]: Good tradeoff IMO

**Pull request merged**